File: kimden/normal/NA/358.cpp
#include <bits/stdc++.h>
#define all(x) (x).begin(), (x).end()
#define rall(x) (x).rbegin(), (x).rend()
using namespace std;
using ll = long long;
using ld = long double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;

ll n, xx, yy;

vector<pll> c;

int solve(vector<ll>& a, vector<ll>& b, vector<ll>& res1, vector<ll>& res2){
    vector<ll> sth(n, -1);
    for(int i = 0; i < n; ++i){
        if(a[i] <= i + 1){
            sth[i] = i;
        }else if(i){
            sth[i] = sth[i - 1];
        }
    }
    for(int j = 1; j < n; ++j){
        ll val = j + 1 - b[j];
        int L = 0;
        int R = j;
        int mid;
        if(a[0] > val){
            continue;
        }
        while(R - L > 1){
            mid = (R + L) / 2;
            if(a[mid] > val){
                R = mid;
            }else{
                L = mid;
            }
        }
        if(sth[L] != -1){
            int ii = sth[L];
            int jj = j;
            int cnt1 = a[ii];
            int cnt2 = b[jj];
            for(int i = 0; i <= jj; ++i){
                if(cnt1){
                    res1.push_back(c[i].second);
                    --cnt1;
                }else{
                    res2.push_back(c[i].second);
                    --cnt2;
                }
            }
            return 1;
        }
    }
    return 0;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin >> n >> xx >> yy;
    vector<ll> a, b;
    c.resize(n);
    a.resize(n);
    b.resize(n);
    for(int i = 0; i < n; ++i){
        cin >> c[i].first;
        c[i].second = i;
    }
    sort(rall(c));
    for(int i = 0; i < n; ++i){
        a[i] = (xx + c[i].first - 1) / c[i].first;
        b[i] = (yy + c[i].first - 1) / c[i].first;
    }
    vector<ll> res1, res2;
    if(solve(a, b, res1, res2)){
        cout << "Yes" << endl;
        cout << res1.size() << " " << res2.size() << endl;
        for(int i = 0; i < res1.size(); ++i){
            if(i){
                cout << " ";
            }
            cout << res1[i] + 1;
        }
        cout << endl;
        for(int i = 0; i < res2.size(); ++i){
            if(i){
                cout << " ";
            }
            cout << res2[i] + 1;
        }
        cout << endl;
    }else if(solve(b, a, res2, res1)){
        cout << "Yes" << endl;
        cout << res1.size() << " " << res2.size() << endl;
        for(int i = 0; i < res1.size(); ++i){
            if(i){
                cout << " ";
            }
            cout << res1[i] + 1;
        }
        cout << endl;
        for(int i = 0; i < res2.size(); ++i){
            if(i){
                cout << " ";
            }
            cout << res2[i] + 1;
        }
        cout << endl;
    }else{
        cout << "No" << endl;
    }
    return 0;
}

File: eddy1021/normal/NA/358.py
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
const int N=55;
int n, a[N], b[N];
vector<pair<LL,LL>> v;
#define P first
#define C second
const LL inf=1e18;
LL dp[N][N][N], vl[N];
bool good(LL goal){
  for(int i=0; i<N; i++)
    for(int j=0; j<N; j++)
      for(int k=0; k<N; k++)
        dp[i][j][k]=inf;
  dp[0][0][0]=0;
  for(int i=0; i<n; i++){
    LL g=v[i].P-goal*v[i].C;
    for(int jj=0; jj<=i; jj++)
      for(int kk=0; jj+kk<=i; kk++){
        if(dp[i][jj][kk] == inf) continue;
        int nj=jj, nk=kk;
        if(i==0 or v[i].P!=v[i-1].P)
          nj+=nk, nk=0;
        dp[i+1][nj][nk+1]=min(dp[i+1][nj][nk+1], dp[i][jj][kk]+g);
        if(nj)
          dp[i+1][nj-1][nk]=min(dp[i+1][nj-1][nk], dp[i][jj][kk]);
      }
  }
  LL ans=inf;
  for(int i=0; i<N; i++)
    for(int j=0; j<N; j++)
      ans=min(ans, dp[n][i][j]);
  return ans<=0;
}
int main(){
  scanf("%d", &n);
  for(int i=0; i<n; i++)
    scanf("%d", &a[i]);
  for(int i=0; i<n; i++)
    scanf("%d", &b[i]);
  for(int i=0; i<n; i++)
    v.push_back({a[i]*1000LL, b[i]});
  sort(v.begin(), v.end(),
       [&](pair<LL,LL>& p1, pair<LL,LL>& p2){
        if(p1.P != p2.P) return p1.P > p2.P;
        return p1.C > p2.C;
       });
  LL bl=0, br=1e12, ans=1;
  while(bl <= br){
    LL bmid=(bl+br)>>1;
    if(good(bmid)) br=bmid-1, ans=bmid;
    else bl=bmid+1;
  }
  printf("%lld\n", ans);
}